File: contract/src/main/java/rrlane/leasing/core/controller/CustomerController.java
package rrlane.leasing.core.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;
import rrlane.leasing.contract.dto.CustomerDTO;
import rrlane.leasing.core.service.CustomerService;

/**
 * Note for readers:
 * The API methods have not been designed in old (more verbose) fashion.
 * Kindly read through the code closely to appreciate the compact yet extensible approach well.
 */
@RestController
@Slf4j
@RequiredArgsConstructor
@RequestMapping(value = "/api/v1/customers")
public class CustomerController {
    private final CustomerService customerService;

    // to create a customer with POST and also to edit a customer with PATCH
    @RequestMapping(method = {RequestMethod.POST, RequestMethod.PATCH})
    @CrossOrigin(origins = "http://localhost:4200/", maxAge = 3600)
    public ResponseEntity<CustomerDTO> uploadCustomer(@RequestBody CustomerDTO customerDTO) {
        log.info("Uploading customer details to server...");
        HttpStatus responseStatus = HttpStatus.OK; // for update
        String responseFromService = customerService.saveCustomer(customerDTO);
        if (responseFromService.contains("created")) {
            responseStatus = HttpStatus.CREATED; // for new customer record creation
        }
        log.info(responseFromService);
        return new ResponseEntity<>(customerDTO, responseStatus);
    }

    // to retrieve the details of a customer with matching inputs (search criteria)
    @GetMapping()
    public ResponseEntity<CustomerDTO> getCustomer(@RequestBody CustomerDTO customerDTO) {
        log.info("Searching for customer details on server...");
        CustomerDTO retrievedCustomerDTO = customerService.searchForCustomer(customerDTO);
        if (null == retrievedCustomerDTO) {
            log.error("Customer details could not be found on server.");
            return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
        }
        log.info("Found the customer details with name (as per records): " + retrievedCustomerDTO.getName());
        return new ResponseEntity<>(retrievedCustomerDTO, HttpStatus.OK);
    }
}

File: contractapi/src/main/kotlin/de/leasing/contract/api/controller/CustomerController.kt
package de.leasing.contract.api.controller

import de.leasing.contract.api.commons.Constants
import de.leasing.contract.api.entity.dto.CustomerReq
import de.leasing.contract.api.entity.dto.CustomerResp
import de.leasing.contract.api.service.CustomerService
import lombok.extern.slf4j.Slf4j
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.util.UriComponentsBuilder
import java.util.UUID

@RestController @RequestMapping("/api/v2/customers/")
@Slf4j
class CustomerController(private val customerService: CustomerService) {
    private val log: Logger = LoggerFactory.getLogger(CustomerController::class.java)

    @PostMapping
    fun addCustomer(
        @RequestBody customerReq: CustomerReq,
        uriBuilder: UriComponentsBuilder
    ): ResponseEntity<CustomerResp> {
        log.info("saving customer details on server...")
        val newCustomerRecord: CustomerResp = customerService.saveCustomerRecord(customerReq)
        log.info(Constants.CUSTOMER_CREATED)

        val location = uriBuilder.path("/api/v2/customers/{id}")
            .buildAndExpand(newCustomerRecord.id).toUri()
        return ResponseEntity.created(location).body(newCustomerRecord)
    }

    @GetMapping("{id}")
    fun viewCustomerById(
        @PathVariable id: UUID
    ) : ResponseEntity<CustomerResp>{
        log.info("Fetching the customer record with the input ID...")
        return ResponseEntity.ok(customerService.getCustomerById(id))
    }

    @GetMapping
    fun viewAllCustomers(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int,
        @RequestParam(required = false) sortBy: String?
    ) : ResponseEntity<Iterable<CustomerResp>>{
        log.info("Fetching the list of all the customers...")
        return  if (null == sortBy) ResponseEntity.ok(customerService.listAllCustomerRecords(page, size))
        else ResponseEntity.ok(customerService.listAllCustomerRecords(sortBy))
    }
}